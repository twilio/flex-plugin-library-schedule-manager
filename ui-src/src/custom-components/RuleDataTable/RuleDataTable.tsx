import React, { useEffect, useState } from 'react';
import { ColumnDefinition, DataTable } from '@twilio/flex-ui';
import { Button } from '@twilio-paste/core/button';
import { Box } from '@twilio-paste/core/box';
import { PlusIcon } from '@twilio-paste/icons/cjs/PlusIcon';
import { RRule } from 'rrule';

import RuleEditor from '../RuleEditor/RuleEditor';
import { Rule, Schedule } from '../../types/schedule-manager';
import ScheduleManagerStrings, { StringTemplates } from '../../flex-hooks/strings/ScheduleManager';
import RRuleLanguage from '../../utils/RRuleLanguage';
import { TableContainer } from '../ScheduleView/ScheduleViewStyles';
import SchedulesAction from '../common/SchedulesAction';
import { Badge, Flex, Toaster, useToaster } from '@twilio-paste/core';
import { StatusBadge } from '@twilio-paste/status';
import AlertBox from '../common/AlertBox';
import { updateRuleData, referencedSchedules } from '../../utils/schedule-manager';

interface OwnProps {
  isLoading: boolean;
  rules: Rule[];
  schedules: Schedule[];
  updateRules: (rules: Rule[]) => void;
}

const RuleDataTable = (props: OwnProps) => {
  const toaster = useToaster();
  const [showPanel, setShowPanel] = useState(false);
  const [selectedRule, setSelectedRule] = useState(null as Rule | null);
  // const [openIndexNext, setOpenIndexNext] = useState(null as number | null);
  const [deleteRule, setDeleteRule] = useState(null as Rule | null);
  const [copyRule, setCopyRule] = useState(false);


  useEffect(() => {
    if (selectedRule !== null) {
      setShowPanel(true);
    }
  }, [selectedRule]);

  const createRuleClick = () => {
    setSelectedRule(null);
    setShowPanel(true);
  };

  const onPanelClosed = () => {
    setShowPanel(false);
    setSelectedRule(null);
  };

  const onEditOrClone = (item: Rule, type: string) => {
    if (type == 'copy') {
      setCopyRule(true);
    }
    setSelectedRule(item);
  };

  const onUpdateRule = (newRules: Rule[]) => {

    props.updateRules(newRules);
    document.querySelector('#rule-data-table-root')?.scrollIntoView({ behavior: 'smooth' });
    setShowPanel(false);
    setSelectedRule(null);
    setCopyRule(false);

  };

  const getRuleRecurrence = (rule: Rule): string => {
    let recStr = ScheduleManagerStrings[StringTemplates.ONE_TIME];

    if (rule.dateRRule) {
      const freq = RRule.fromString(rule.dateRRule).options.freq;
      switch (freq) {
        case 0:
          recStr = 'Every year';
          break;
        case 1:
          recStr = 'Every month';
          break;
        case 2:
          recStr = 'Every week';
          break;
        case 3:
          recStr = 'Daily';
          break;
        default:
          break;
      }
      return recStr;
    }
    return recStr;
  };

  const getRuleTime = (rule: Rule): any => {
    let timeStr = ScheduleManagerStrings[StringTemplates.ANY_TIME];
    if (rule.isOpen) {
      if (rule.startTime) {
        timeStr = rule.startTime;
      }

      if (rule.endTime) {
        timeStr += ` - ${rule.endTime}`;
      }

      return (
        <Badge as="span" variant="success">
          {`Open ${timeStr}`}
        </Badge>
      );
    } else {
      timeStr = ScheduleManagerStrings[StringTemplates.CLOSEDALLDAY];
      return (
        <Badge as="span" variant="warning">
          {timeStr}
        </Badge>
      );
    }
  };

  const getRuleDate = (rule: Rule): string => {
    let dateStr = ScheduleManagerStrings[StringTemplates.ANY_DAY];

    if (rule.startDate && rule.endDate && rule.startDate == rule.endDate) {
      dateStr = `${rule.startDate}`;
    } else {
      dateStr = '';

      if (rule.dateRRule) {
        const text = RRule.fromString(rule.dateRRule).toText(undefined, RRuleLanguage);
        dateStr += text.charAt(0).toUpperCase() + text.slice(1);
      }

      if (dateStr && (rule.startDate || rule.endDate)) {
        dateStr += ', ';
      }

      if (rule.startDate) {
        dateStr += `${ScheduleManagerStrings[StringTemplates.DATE_FROM]} ${rule.startDate}`;
      }

      if (rule.endDate) {
        if (dateStr) {
          dateStr += ' ';
        }
        dateStr += `${ScheduleManagerStrings[StringTemplates.DATE_UNTIL]} ${rule.endDate}`;
      }
    }

    return dateStr;
  };

  const handleDeleteConfirmSubmit = (): void => {
    if (!deleteRule) {
      return;
    }

    const refSchedules = referencedSchedules(props.schedules, deleteRule as Rule);
    if (refSchedules.length > 0) {
      toaster.push({
        message: ScheduleManagerStrings[StringTemplates.ERROR_RULE_REFERENCED] + ' ' + refSchedules.join(', '),
        variant: 'error',
        dismissAfter: 4000,
      });
      return;
    }
    const newScheduleData = updateRuleData(null, deleteRule);
    onUpdateRule(newScheduleData);
    setDeleteRule(null);
    setSelectedRule(null);
  };

  return (
    <>
      <div id="rule-data-table-root">
        <Box padding="space60" display={'flex'} justifyContent={'flex-end'}>
          <Button variant="primary" data-testid="create-rule-btn" disabled={props.isLoading} onClick={createRuleClick}>
            <PlusIcon decorative />
            {ScheduleManagerStrings[StringTemplates.CREATE_RULE_BUTTON]}
          </Button>
        </Box>
        <TableContainer>
          <DataTable
            items={props.rules}
            isLoading={props.isLoading}
            // onRowClick={onRowClick}
            defaultSortColumn="name-column"
          >
            <ColumnDefinition
              key="name-column"
              header={ScheduleManagerStrings[StringTemplates.NAME]}
              sortDirection="asc"
              sortingFn={(a: Rule, b: Rule) => (a.name > b.name ? 1 : -1)}
              content={(item: Rule) => <span id={item.name}>{item.name}</span>}
            />
            <ColumnDefinition
              key="time-column"
              header={ScheduleManagerStrings[StringTemplates.TIME]}
              sortingFn={(a: Rule, b: Rule) => (getRuleTime(a) > getRuleTime(b) ? 1 : -1)}
              content={(item: Rule) => getRuleTime(item)}
            />
            <ColumnDefinition
              key="date-column"
              header={ScheduleManagerStrings[StringTemplates.DATE]}
              sortingFn={(a: Rule, b: Rule) => (getRuleDate(a) > getRuleDate(b) ? 1 : -1)}
              content={(item: Rule) => <span>{getRuleDate(item)}</span>}
            />
            <ColumnDefinition
              key="recurrence-column"
              header={ScheduleManagerStrings[StringTemplates.DATE]}
              sortingFn={(a: Rule, b: Rule) => (getRuleRecurrence(a) > getRuleRecurrence(b) ? 1 : -1)}
              content={(item: Rule) => <span>{getRuleRecurrence(item)}</span>}
            />
            <ColumnDefinition
              key="publish-status-column"
              header={ScheduleManagerStrings[StringTemplates.COLUMN_PUBLISHSTATUS]}
              // sortingFn={(a: Rule, b: Rule) => (a.manualClose ? 1 : -1)}
              content={(item: Rule) => (
                <StatusBadge as="span" variant="ProcessSuccess">
                  Published
                </StatusBadge>
              )}
            />
            <ColumnDefinition
              key="actions-column"
              header=""
              content={(item: Rule) => (
                <SchedulesAction
                  onCopy={() => {
                    onEditOrClone(item, 'copy');
                  }}
                  deleteDisabled={false}
                  editDisabled={false}
                  copyDisabled={false}
                  onDelete={() => setDeleteRule(item)}
                  onEdit={() => onEditOrClone(item, 'edit')}
                />
              )}
            />
          </DataTable>
        </TableContainer>
      </div>
      <Toaster {...toaster} />
      <RuleEditor
        onPanelClosed={onPanelClosed}
        showPanel={showPanel}
        copy={copyRule}
        schedules={props.schedules}
        onDelete={() => setDeleteRule(selectedRule)}
        selectedRule={selectedRule}
        onUpdateRule={onUpdateRule}
      />
      <AlertBox
        handleClose={() => setDeleteRule(null)}
        isOpen={!!deleteRule}
        handleSubmit={() => handleDeleteConfirmSubmit()}
        title={ScheduleManagerStrings[StringTemplates.DELETE_CONFIRM_RULE_TITLE]}
      >
        <Flex>
          <Box as="p">{ScheduleManagerStrings[StringTemplates.DELETE_CONFIRM_RULE]}</Box>
          <Box as="strong" fontWeight={'fontWeightSemibold'} marginLeft={'space10'}>
            {deleteRule?.name}
          </Box>
        </Flex>
      </AlertBox>
    </>
  );
};

export default RuleDataTable;
