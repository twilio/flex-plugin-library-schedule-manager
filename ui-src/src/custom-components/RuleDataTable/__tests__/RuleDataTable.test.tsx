import React from 'react';
import { render } from '@testing-library/react';
import RuleDataTable from '../RuleDataTable';

describe('<RuleDataTable>', () => {
  it('should render component', () => {
    expect(
      render(
        <RuleDataTable
          isLoading={true}
          rules={[]}
          schedules={[]}
          updateRules={() => {}}
          toaster={{ push: () => {}, pop: () => {} } as any}
        />,
      ),
    ).toMatchSnapshot();
  });
  it('should render component with props isLoading true and create rule button disabled', () => {
    const { getByTestId } = render(
      <RuleDataTable
        isLoading={true}
        rules={[]}
        schedules={[]}
        updateRules={() => {}}
        toaster={{ push: () => {}, pop: () => {} } as any}
      />,
    );
    expect(getByTestId('create-rule-btn').disabled).toBeTruthy();
  });
});
