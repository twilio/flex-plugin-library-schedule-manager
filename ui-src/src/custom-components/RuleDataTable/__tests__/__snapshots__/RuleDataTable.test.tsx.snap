// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RuleDataTable> should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="rule-data-table-root"
      >
        <div
          class="css-cuork0"
          data-paste-element="BOX"
        >
          <button
            aria-busy="false"
            class="css-eeqvux"
            data-paste-element="BUTTON"
            data-testid="create-rule-btn"
            disabled=""
            type="button"
          >
            <span
              class="css-z6jp9v"
              data-paste-element="BOX"
            >
              <span
                class="css-144s81u"
                data-paste-element="ICON"
              >
                <svg
                  aria-hidden="true"
                  aria-labelledby="PlusIcon-1"
                  height="100%"
                  role="img"
                  viewBox="0 0 20 20"
                  width="100%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15.043 9.043h-4.086V4.957a.958.958 0 00-1.914 0v4.086H4.957a.958.958 0 000 1.914h4.086v4.086a.958.958 0 001.914 0v-4.086h4.086a.958.958 0 000-1.914z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              Create new rule
            </span>
          </button>
        </div>
        <div
          class="css-1fha43i"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="rule-data-table-root"
    >
      <div
        class="css-cuork0"
        data-paste-element="BOX"
      >
        <button
          aria-busy="false"
          class="css-eeqvux"
          data-paste-element="BUTTON"
          data-testid="create-rule-btn"
          disabled=""
          type="button"
        >
          <span
            class="css-z6jp9v"
            data-paste-element="BOX"
          >
            <span
              class="css-144s81u"
              data-paste-element="ICON"
            >
              <svg
                aria-hidden="true"
                aria-labelledby="PlusIcon-1"
                height="100%"
                role="img"
                viewBox="0 0 20 20"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.043 9.043h-4.086V4.957a.958.958 0 00-1.914 0v4.086H4.957a.958.958 0 000 1.914h4.086v4.086a.958.958 0 001.914 0v-4.086h4.086a.958.958 0 000-1.914z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            Create new rule
          </span>
        </button>
      </div>
      <div
        class="css-1fha43i"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
